File: go/test/cart_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/fake-store-sdk"
	"github.com/voxgig-sdk/fake-store-sdk/core"

	vs "github.com/voxgig/struct"
)

func TestCartEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Cart(nil)
		if ent == nil {
			t.Fatal("expected non-nil CartEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := cartBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "list", "update", "load", "remove"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "cart." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set FAKESTORE_TEST_CART_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		cartRef01Ent := client.Cart(nil)
		cartRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath([]any{"new", "cart"}, setup.data), "cart_ref01"))

		cartRef01DataResult, err := cartRef01Ent.Create(cartRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		cartRef01Data = core.ToMapAny(cartRef01DataResult)
		if cartRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}
		if cartRef01Data["id"] == nil {
			t.Fatal("expected created entity to have an id")
		}

		// LIST
		cartRef01Match := map[string]any{}

		cartRef01ListResult, err := cartRef01Ent.List(cartRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		cartRef01List, cartRef01ListOk := cartRef01ListResult.([]any)
		if !cartRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", cartRef01ListResult)
		}

		foundItem := vs.Select(entityListToData(cartRef01List), map[string]any{"id": cartRef01Data["id"]})
		if vs.IsEmpty(foundItem) {
			t.Fatal("expected to find created entity in list")
		}

		// UPDATE
		cartRef01DataUp0Up := map[string]any{
			"id": cartRef01Data["id"],
		}

		cartRef01ResdataUp0Result, err := cartRef01Ent.Update(cartRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		cartRef01ResdataUp0 := core.ToMapAny(cartRef01ResdataUp0Result)
		if cartRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if cartRef01ResdataUp0["id"] != cartRef01DataUp0Up["id"] {
			t.Fatal("expected update result id to match")
		}

		// LOAD
		cartRef01MatchDt0 := map[string]any{
			"id": cartRef01Data["id"],
		}
		cartRef01DataDt0Loaded, err := cartRef01Ent.Load(cartRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		cartRef01DataDt0LoadResult := core.ToMapAny(cartRef01DataDt0Loaded)
		if cartRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if cartRef01DataDt0LoadResult["id"] != cartRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

		// REMOVE
		cartRef01MatchRm0 := map[string]any{
			"id": cartRef01Data["id"],
		}
		_, err = cartRef01Ent.Remove(cartRef01MatchRm0, nil)
		if err != nil {
			t.Fatalf("remove failed: %v", err)
		}

		// LIST
		cartRef01MatchRt0 := map[string]any{}

		cartRef01ListRt0Result, err := cartRef01Ent.List(cartRef01MatchRt0, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		cartRef01ListRt0, cartRef01ListRt0Ok := cartRef01ListRt0Result.([]any)
		if !cartRef01ListRt0Ok {
			t.Fatalf("expected list result to be an array, got %T", cartRef01ListRt0Result)
		}

		notFoundItem := vs.Select(entityListToData(cartRef01ListRt0), map[string]any{"id": cartRef01Data["id"]})
		if !vs.IsEmpty(notFoundItem) {
			t.Fatal("expected removed entity to not be in list")
		}

	})
}

func cartBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "cart", "CartTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read cart test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse cart test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"cart01", "cart02", "cart03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("FAKESTORE_TEST_CART_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"FAKESTORE_TEST_CART_ENTID": idmap,
		"FAKESTORE_TEST_LIVE":      "FALSE",
		"FAKESTORE_TEST_EXPLAIN":   "FALSE",
		"FAKESTORE_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["FAKESTORE_TEST_CART_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["FAKESTORE_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["FAKESTORE_APIKEY"],
			},
			extra,
		})
		client = sdk.NewFakeStoreSDK(core.ToMapAny(mergedOpts))
	}

	live := env["FAKESTORE_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["FAKESTORE_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: lua/test/cart_entity_test.lua
-- Cart entity test

local json = require("dkjson")
local vs = require("utility.struct.struct")
local sdk = require("fake-store_sdk")
local helpers = require("core.helpers")
local runner = require("test.runner")

local _test_dir = debug.getinfo(1, "S").source:match("^@(.+/)")  or "./"

describe("CartEntity", function()
  it("should create instance", function()
    local testsdk = sdk.test(nil, nil)
    local ent = testsdk:Cart(nil)
    assert.is_not_nil(ent)
  end)

  it("should run basic flow", function()
    local setup = cart_basic_setup(nil)
    -- Per-op sdk-test-control.json skip.
    local _live = setup.live or false
    for _, _op in ipairs({"create", "list", "update", "load", "remove"}) do
      local _should_skip, _reason = runner.is_control_skipped("entityOp", "cart." .. _op, _live and "live" or "unit")
      if _should_skip then
        pending(_reason or "skipped via sdk-test-control.json")
        return
      end
    end
    -- The basic flow consumes synthetic IDs from the fixture. In live mode
    -- without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup.synthetic_only then
      pending("live entity test uses synthetic IDs from fixture — set FAKESTORE_TEST_CART_ENTID JSON to run live")
      return
    end
    local client = setup.client

    -- CREATE
    local cart_ref01_ent = client:Cart(nil)
    local cart_ref01_data = helpers.to_map(vs.getprop(
      vs.getpath(setup.data, "new.cart"), "cart_ref01"))

    local cart_ref01_data_result, err = cart_ref01_ent:create(cart_ref01_data, nil)
    assert.is_nil(err)
    cart_ref01_data = helpers.to_map(cart_ref01_data_result)
    assert.is_not_nil(cart_ref01_data)
    assert.is_not_nil(cart_ref01_data["id"])

    -- LIST
    local cart_ref01_match = {}

    local cart_ref01_list_result, err = cart_ref01_ent:list(cart_ref01_match, nil)
    assert.is_nil(err)
    assert.is_table(cart_ref01_list_result)

    local found_item = vs.select(
      runner.entity_list_to_data(cart_ref01_list_result),
      { id = cart_ref01_data["id"] })
    assert.is_false(vs.isempty(found_item))

    -- UPDATE
    local cart_ref01_data_up0_up = {
      id = cart_ref01_data["id"],
    }

    local cart_ref01_resdata_up0_result, err = cart_ref01_ent:update(cart_ref01_data_up0_up, nil)
    assert.is_nil(err)
    local cart_ref01_resdata_up0 = helpers.to_map(cart_ref01_resdata_up0_result)
    assert.is_not_nil(cart_ref01_resdata_up0)
    assert.are.equal(cart_ref01_resdata_up0["id"], cart_ref01_data_up0_up["id"])

    -- LOAD
    local cart_ref01_match_dt0 = {
      id = cart_ref01_data["id"],
    }
    local cart_ref01_data_dt0_loaded, err = cart_ref01_ent:load(cart_ref01_match_dt0, nil)
    assert.is_nil(err)
    local cart_ref01_data_dt0_load_result = helpers.to_map(cart_ref01_data_dt0_loaded)
    assert.is_not_nil(cart_ref01_data_dt0_load_result)
    assert.are.equal(cart_ref01_data_dt0_load_result["id"], cart_ref01_data["id"])

    -- REMOVE
    local cart_ref01_match_rm0 = {
      id = cart_ref01_data["id"],
    }
    local _, err = cart_ref01_ent:remove(cart_ref01_match_rm0, nil)
    assert.is_nil(err)

    -- LIST
    local cart_ref01_match_rt0 = {}

    local cart_ref01_list_rt0_result, err = cart_ref01_ent:list(cart_ref01_match_rt0, nil)
    assert.is_nil(err)
    assert.is_table(cart_ref01_list_rt0_result)

    local not_found_item = vs.select(
      runner.entity_list_to_data(cart_ref01_list_rt0_result),
      { id = cart_ref01_data["id"] })
    assert.is_true(vs.isempty(not_found_item))

  end)
end)

function cart_basic_setup(extra)
  runner.load_env_local()

  local entity_data_file = _test_dir .. "../../.sdk/test/entity/cart/CartTestData.json"
  local f = io.open(entity_data_file, "r")
  if f == nil then
    error("failed to read cart test data: " .. entity_data_file)
  end
  local entity_data_source = f:read("*a")
  f:close()

  local entity_data = json.decode(entity_data_source)

  local options = {}
  options["entity"] = entity_data["existing"]

  local client = sdk.test(options, extra)

  -- Generate idmap via transform.
  local idmap = vs.transform(
    { "cart01", "cart02", "cart03" },
    {
      ["`$PACK`"] = { "", {
        ["`$KEY`"] = "`$COPY`",
        ["`$VAL`"] = { "`$FORMAT`", "upper", "`$COPY`" },
      }},
    }
  )

  -- Detect ENTID env override before envOverride consumes it. When live
  -- mode is on without a real override, the basic test runs against synthetic
  -- IDs from the fixture and 4xx's. Surface this so the test can skip.
  local entid_env_raw = os.getenv("FAKESTORE_TEST_CART_ENTID")
  local idmap_overridden = entid_env_raw ~= nil and entid_env_raw:match("^%s*{") ~= nil

  local env = runner.env_override({
    ["FAKESTORE_TEST_CART_ENTID"] = idmap,
    ["FAKESTORE_TEST_LIVE"] = "FALSE",
    ["FAKESTORE_TEST_EXPLAIN"] = "FALSE",
  })

  local idmap_resolved = helpers.to_map(
    env["FAKESTORE_TEST_CART_ENTID"])
  if idmap_resolved == nil then
    idmap_resolved = helpers.to_map(idmap)
  end

  if env["FAKESTORE_TEST_LIVE"] == "TRUE" then
    local merged_opts = vs.merge({
      {
      },
      extra or {},
    })
    client = sdk.new(helpers.to_map(merged_opts))
  end

  local live = env["FAKESTORE_TEST_LIVE"] == "TRUE"
  return {
    client = client,
    data = entity_data,
    idmap = idmap_resolved,
    env = env,
    explain = env["FAKESTORE_TEST_EXPLAIN"] == "TRUE",
    live = live,
    synthetic_only = live and not idmap_overridden,
    now = os.time() * 1000,
  }
end
